### DISCUSSION TITLE
Union gives errors on last type only

### DISCUSSION THREAD

**Discussion by rsinohara** [colinhacks/zod#590] [Q&A]
Let's say I have a Union of two types. I expect my value to match the first type, but something is wrong.
The problem is Zod only reports the error on the second type, so that doesn't help me to debug what is wrong in my first type.

Am I missing something? Right now (just starting) I remove the union just for debugging, but that will not be possible at scale.

**Comment by scotttrinh** [comment]
@rsinohara it's a little confusing because it's an array of arrays, but you should see something from your target schema reporting errors in the list of issues. feel free to post an example here!

**Reply by rsinohara** [reply]
Ah, so the first type error should also be visible? That helps.

Is there anyway to format this error output while parsing at all, or this array of error is the only way?

**Reply by scotttrinh** [reply]
There is some documentation here: https://github.com/colinhacks/zod/blob/master/ERROR_HANDLING.md and specifically the `flatten` method mentioned at the end might be of interest in your case.

In the longer term, I'm also planning on tackling some better support for discriminated unions so that instead of reporting all of the union errors, you specify a discriminant and then it checks only against the matching schema (if it exists), which should help a bit here. Doesn't help with non-discriminated unions or unions of primitives, etc.

**Reply by rsinohara** [reply]
Interesting. Did anyone explore having a different output, like an 'annotated JSON', something similar to the current error but like this (besides the comments, everything is the original value being parsed):

```
{
  name: "Renato",
  lastName: "Sinohara" // Zod schema error:  "code": "invalid_type","expected": "null",
...
}
```

**Reply by scotttrinh** [reply]
I think that would be tough unless we had a way of knowing which of the schemas was your intended target (which is the purpose of the exploration I mentioned above). Otherwise, what would this output:

```ts
const schema = z.union([
  z.object({ a: z.string() }),
  z.object({ b: z.string() }),
]);

schema.parse({ c: 42 });
{
  c: 42 // what goes here? what happened to a and b?
}
```

Compare with (this is not a real thing yet, so just bare with me):

```ts
const schema = z.discriminatedUnion([
  z.object({ type: z.literal("a"), a: z.string() }),
  z.object({ type: z.literal("b"), b: z.string() }),
], "type");

schema.parse({ type: "a", c: 42 }); // we know you meant the a schema
schema.parse({ type: "c", c: 42 }); // we know this union does not contain this member
```

**Reply by colinhacks** [reply]
@rsinohara You can get the errors in a nested/structured form if you call `error.format()`, but it's not the "annotation JSON" concept you describe (which is difficult for the reasons Scott describes).

**Reply by rsinohara** [reply]
I see, it does get complex.
I was only thinking of the errors I encountered so far (and I am parsing many already existing objects).

But just out of curiosity. To me the annotation there wouldn't be con C, but on the object itself, something line 
```
{
  c: 42 
} // Could not match to any union type
```
I will get back to thinking about this once I learn more about what types of errors can happen.

**Reply by scotttrinh** [reply]
I guess the point is that error doesn't tell you anything. Why didn't it match? It's about as helpful as `result.success === false`. Is it just missing an `a` or a `b` property?

**Reply by rsinohara** [reply]
True.
My main idea was to place the error inside the evaluated object. Mostly because in a large object, even the non specific error message is better close to where it happens. For sure improving the intelligence around the errors would also be great.

Thanks for looking into it. I'll keep it in mind and see if something comes out of it.

**Reply by colinhacks** [reply]
There is detailed error reporting with unions that describes exactly why each element failed to parse the input. @rsinohara you'll have to provide a code example if you still think something is behaving unexpectedly.

```ts
const schema = z.union([
  z.object({ a: z.string() }),
  z.object({ b: z.string() }),
]);

console.log(schema.safeParse({ c: 42 }));
/*
ZodError: [
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "undefined",
            "path": [
              "a"
            ],
            "message": "Required"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "undefined",
            "path": [
              "b"
            ],
            "message": "Required"
          }
        ],
        "name": "ZodError"
      }
    ],
    "path": [],
    "message": "Invalid input"
  }
]
*/
```

**Reply by rsinohara** [reply]
Hi Colin, that is what I see, all good.
I was just asking if there was any other way to format the error. To me, they would be more useful if shown as 'annotation' in the original object being parsed, so I wanted to make sure there were no other formatting options available.

I will be doing a lot of huge object schemas, so if there is a chance this could work let me know and I can play with it.

**Comment by rsinohara** [comment]
Actually @colinhacks I have one specific case where I only get errors on the second union type.

Is this behavior known or expected?

This schema is defined by `export const ProductRowListSchema = z.array(z.array(CiSchema.or(ProductSchema)));`
Funny enough, changing the order of the union doesn't help.
This particular element was expected to be a 'product'. It fails because a nested schema should be nullable. But I only ever see errors for the CISchema part.
```
{
    "issues": [
        {
            "code": "invalid_union",
            "unionErrors": [
                {
                    "issues": [
                        {
                            "code": "invalid_type",
                            "expected": "content_infusion",
                            "received": "product",
                            "path": [0, 0, "type"],
                            "message": "Expected content_infusion, received product"
                        },
                        {
                            "code": "invalid_type",
                            "expected": "string",
                            "received": "undefined",
                            "path": [0, 0, "id"],
                            "message": "Required"
                        },
                        {
                            "code": "invalid_type",
                            "expected": "ci_collage",
                            "received": null,
                            "path": [0, 0, "subtype"],
                            "message": "Expected ci_collage, received null"
                        },
                        {
                            "code": "invalid_type",
                            "expected": "object",
                            "received": "undefined",
                            "path": [0, 0, "image"],
                            "message": "Required"
                        }
                    ],
                    "name": "ZodError"
                },
                {
                    "issues": [
                        {
                            "code": "invalid_type",
                            "expected": 3,
                            "received": 1,
                            "path": [0, 0, "width"],
                            "message": "Expected 3, received 1"
                        },
                        {
                            "code": "invalid_type",
                            "expected": "string",
                            "received": "object",
                            "path": [0, 0, "content"],
                            "message": "Expected string, received object"
                        },
                        {
                            "code": "invalid_type",
                            "expected": "content_infusion",
                            "received": "product",
                            "path": [0, 0, "type"],
                            "message": "Expected content_infusion, received product"
                        },
                        {
                            "code": "invalid_type",
                            "expected": "string",
                            "received": "undefined",
                            "path": [0, 0, "id"],
                            "message": "Required"
                        },
                        {
                            "code": "invalid_type",
                            "expected": "ci_image",
                            "received": null,
                            "path": [0, 0, "subtype"],
                            "message": "Expected ci_image, received null"
                        },
                        {
                            "code": "invalid_type",
                            "expected": "object",
                            "received": "undefined",
                            "path": [0, 0, "image"],
                            "message": "Required"
                        }
                    ],
                    "name": "ZodError"
                }
            ],
            "path": [0, 0],
            "message": "Invalid input"
        }
    ],
    "name": "ZodError"
}

```

**Reply by colinhacks** [reply]
Looks like `unionErrors` contains two elements, the first corresponds to CiSchema and the second corresponds to ProductSchema.

**Reply by rsinohara** [reply]
The two errors are from content infusion schema error, there are just two of them. CiSchema is itself an union. The error for product is really not there.

This is consistent, I am constantly running the same check but filtering only products, then I get the correct errors.
